openfe/protocols/openmm_utils/multistate_analysis.py

```diff
@@ -0,0 +1,619 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+"""
+Reusable utility methods to analyze results from multistate calculations.
+"""
+from pathlib import Path
+from typing import Union, Optional
+import warnings
+
+import matplotlib.pyplot as plt
+import numpy as np
+import numpy.typing as npt
+from openmmtools import multistate
+from openff.units import unit, ensure_quantity
+
+
+class MultistateEquilFEAnalysis:
+    """
+    A class to generate and plot all necessary analyses for a free energy
+    calculation using a :class:`openmmtools.MultiStateSampler`.
+
+    Currently implemented analyses are:
+      - Decorrelated MBAR analysis of free energies (and associated errors)
+      - Number of equilibration & sampling steps
+      - MBAR overlap matrices
+      - Replica lambda traversal (exchange matrix and timeseries)
+      - Forward and reverse analysis of free energies
+
+    Parameters
+    ----------
+    reporter : openmmtools.MultiStateReporter
+      Reporter for the MultiStateSampler
+    sampling_method : str
+      The sampling method. Expected values are `repex`, `sams`,
+      and `independent`.
+    result_units : unit.Quantity
+      Units to report results in.
+    forward_reverse_samples : int
+      The number of samples to use in the foward and reverse analysis
+      of the free energies. Default 10.
```

Comment:
this signature is confusing, who is supposed to use this class and when is it designed to be used?  Is this during a simulation or after a simulation?  Is this internal code or using facing?  It seems confused between the two.

Rather than accepting a reporter, can we just request the `.nc` path (which we can trivially create the Reporter from).

Having forward/reverse samples defined here doesn't make sense, surely that's an argument to the related method?

Reply:
> this signature is confusing, who is supposed to use this class and when is it designed to be used? Is this during a simulation or after a simulation? Is this internal code or using facing? It seems confused between the two.

Unfortunately we are trying to do both here, there's both a need to have a class that can generate all this information externally (if you just have a netcdf file & checkpoint), and there's a need to have something that you could use both during and after a simulation within a protocol unit.

I agree that it's not the most amazing API, but I also need this **yesterday**.

> Rather than accepting a reporter, can we just request the .nc path (which we can trivially create the Reporter from).

Because of the very flaky way reporters work, this would require us to both force delete caches and close the reporter twice during a protocol unit, which is a real pain. Until we have a bettter upstream way to do this, passing the reporter through here avoids _some_ mess - indeed if you look at the `_analyze` code you'll realise that stuff that _should_ just be one liners called every time you call an attributte don't happen because even then the whole thing falls apart if you query certain things too often (don't ask me why..).

> Having forward/reverse samples defined here doesn't make sense, surely that's an argument to the related method?

I don't know of any other way to pass that through to `_analyze` unless you want to just always use the default value?